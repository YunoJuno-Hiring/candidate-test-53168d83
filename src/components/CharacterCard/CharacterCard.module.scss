@import 'styles/_mixins';

.CharacterCard {
  background-color: white;
  padding: 16px;
  margin: 16px;
  display: flex;
  color: $fontColor;

  @include lg {
    max-width: 40vw;
    min-width: 30vw;
  }
}

.CharacterCard-avatar-container {
  display: flex;
  align-items: center;
  justify-content: center;
}

.CharacterCard-avatar {
  width: 80px;
  border-radius: 50%;
}

.CharacterCard-content-container {
  display: flex;
  flex-direction: column;
  margin: 0px 16px;
  justify-content: space-between;
}

.CharacterCard-content {
  margin: 8px;
}
