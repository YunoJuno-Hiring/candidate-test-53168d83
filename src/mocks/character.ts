import { ICharacter } from '@custom-types/character';

export function generateMockCharacter(overrides?: Partial<ICharacter>): ICharacter {
  const mockCharacter = {
    name: 'Gandalf the Grey',
    category: 'wizard' as const,
    description:
      'A wizard best known among hobbits for his fireworks and mischievous sense of humor, but actually one of the greatest powers of Middle-earth. He reveals the truth about the Ring to Frodo and acts as a guide and counselor.',
    significanceIndex: 1,
    avatar: 'gandalf_the_grey.jpg'
  };

  return { ...mockCharacter, ...overrides };
}
