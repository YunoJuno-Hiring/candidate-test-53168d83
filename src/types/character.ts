import * as strings from '@constants/strings';

const { characterCategoryStrings } = strings;

export type TCharacterCategories =
  (typeof characterCategoryStrings)[keyof typeof characterCategoryStrings];

export interface ICharacter {
  category: TCharacterCategories;
  name: string;
  description: string;
  significanceIndex: number;
  avatar: string;
}
