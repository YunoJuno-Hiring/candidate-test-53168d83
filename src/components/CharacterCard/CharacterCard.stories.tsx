import type { Meta, StoryObj } from '@storybook/react';
import CharacterCard from './CharacterCard';
import { generateMockCharacter } from '@mocks/character';
import { Character } from '@custom-types/character';

const meta: Meta<typeof CharacterCard> = {
  component: CharacterCard,
  tags: ['autodocs']
};

export default meta;

const mockCharacter: Character = generateMockCharacter();

type Story = StoryObj<typeof CharacterCard>;
export const Default: Story = {
  args: {
    characterData: {
      ...mockCharacter
    }
  }
};
