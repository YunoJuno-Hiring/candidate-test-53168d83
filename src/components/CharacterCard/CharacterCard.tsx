import { Character } from '@custom-types/character';
import React from 'react';
import styles from './CharacterCard.module.scss';
import { capitaliseFirstCharacters } from '@utils/stringHelpers';

type CharacterCardProps = {
  characterData: Character;
};

export default function CharacterCard({ characterData }: CharacterCardProps) {
  const avatarUrl = `${process.env.REACT_APP_AVATAR_DIR as string}${characterData.avatar}`;

  return (
    <div className={styles['CharacterCard']}>
      <div className={styles['CharacterCard-avatar-wrapper']}>
        <img className={styles['CharacterCard-avatar']} src={avatarUrl} />
      </div>
      <div className={styles['CharacterCard-content-container']}>
        <p className={styles['CharacterCard-content']}>
          <strong>{characterData.name}</strong>
        </p>
        <p className={styles['CharacterCard-content']}>
          {capitaliseFirstCharacters(characterData.category)}
        </p>
        <p className={styles['CharacterCard-content']}>{characterData.description}</p>
      </div>
    </div>
  );
}
