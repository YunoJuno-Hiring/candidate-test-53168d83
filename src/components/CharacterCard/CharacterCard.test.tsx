import React from 'react';
import { generateMockCharacter } from '@mocks/character';
import CharacterCard from './CharacterCard';
import renderer from 'react-test-renderer';

const mockCharacterData = generateMockCharacter({ category: 'hobbit' });

describe('CharacterCard', () => {
  it('matches the DOM snapshot', () => {
    const component = renderer.create(<CharacterCard characterData={mockCharacterData} />).toJSON();
    expect(component).toMatchSnapshot();
  });
});
