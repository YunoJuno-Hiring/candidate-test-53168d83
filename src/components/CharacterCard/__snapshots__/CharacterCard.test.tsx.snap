// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard matches the DOM snapshot 1`] = `
<div
  className="CharacterCard"
>
  <div
    className="CharacterCard-avatar-wrapper"
  >
    <img
      className="CharacterCard-avatar"
      src="undefinedgandalf_the_grey.jpg"
    />
  </div>
  <div
    className="CharacterCard-content-container"
  >
    <p
      className="CharacterCard-content"
    >
      <strong>
        Gandalf the Grey
      </strong>
    </p>
    <p
      className="CharacterCard-content"
    >
      Hobbit
    </p>
    <p
      className="CharacterCard-content"
    >
      A wizard best known among hobbits for his fireworks and mischievous sense of humor, but actually one of the greatest powers of Middle-earth. He reveals the truth about the Ring to Frodo and acts as a guide and counselor.
    </p>
  </div>
</div>
`;
